src/clients/BundleDataClient/BundleDataClient.ts

```diff
@@ -1067,37 +1047,124 @@ export class BundleDataClient {
                 "Deposit relay hashes should match."
               );
-
-              // slow fill requests for deposits from or to lite chains are considered invalid
-              if (matchedDeposit.fromLiteChain || matchedDeposit.toLiteChain) {
-                return;
-              }
-
               v3RelayHashes[relayDataHash].deposit = matchedDeposit;
 
-              // Note: we don't need to query for a historical fill at this point because a fill
-              // cannot precede a slow fill request and if the fill came after the slow fill request,
-              // we would have seen it already because we would have processed it in the loop above.
               if (
-                // Input and Output tokens must be equivalent on the deposit for this to be slow filled.
-                !this.clients.hubPoolClient.areTokensEquivalent(
-                  matchedDeposit.inputToken,
-                  matchedDeposit.originChainId,
-                  matchedDeposit.outputToken,
-                  matchedDeposit.destinationChainId,
-                  matchedDeposit.quoteBlockNumber
-                ) ||
+                !_canCreateSlowFillLeaf(matchedDeposit) ||
                 // Deposit must not have expired in this bundle.
-                slowFillRequest.fillDeadline < bundleBlockTimestamps[destinationChainId][1]
+                _depositIsExpired(matchedDeposit)
               ) {
-                // TODO: Invalid slow fill request. Maybe worth logging.
                 return;
               }
               validatedBundleSlowFills.push(matchedDeposit);
             }
           }
         );
 
+        // Deposits can be submitted an arbitrary amount of time after matching fills and slow fill requests.
+        // Therefore, let's go through each deposit in this bundle again and check a few things in order:
+        // - Has the deposit been filled ? If so, then we need to issue a relayer refund for
+        //   this "pre-fill" if the fill took place in a previous bundle.
+        // - Or, has the deposit expired in this bundle? If so, then we need to issue an expiry refund.
+        // - And finally, has the deposit been slow filled? If so, then we need to issue a slow fill leaf
+        //   for this "pre-slow-fill-request" if this request took place in a previous bundle.
+
+        // @todo Only start refunding pre-fills and slow fill requests after a config store version is activated. We
+        // should remove this check once we've advanced far beyond the version bump block.
+        await mapAsync(
+          bundleDepositHashes.filter((depositHash) => {
+            const { deposit } = v3RelayHashes[depositHash];
+            return (
+              deposit && deposit.originChainId === originChainId && deposit.destinationChainId === destinationChainId
+            );
+          }),
+          async (depositHash) => {
+            const { deposit, fill, slowFillRequest } = v3RelayHashes[depositHash];
```

Comment:
We don't 
```suggestion
            // We don't need to call verifyFillRepayment() here to replace the fill.relayer because this value should already
            // be overwritten because the deposit and fill both exist.
            const { deposit, fill, slowFillRequest } = v3RelayHashes[depositHash];
```